util/ranger/detacher.go

```diff
@@ -562,26 +604,46 @@ func (d *rangeDetacher) detachDNFCondAndBuildRangeForIndex(condition *expression
 	dnfItems := expression.FlattenDNFConditions(condition)
 	newAccessItems := make([]expression.Expression, 0, len(dnfItems))
 	var totalRanges []*Range
+	columnValues := make([]*valueInfo, len(d.cols))
 	hasResidual := false
-	for _, item := range dnfItems {
+	for i, item := range dnfItems {
 		if sf, ok := item.(*expression.ScalarFunction); ok && sf.FuncName.L == ast.LogicAnd {
 			cnfItems := expression.FlattenCNFConditions(sf)
 			var accesses, filters []expression.Expression
 			res, err := d.detachCNFCondAndBuildRangeForIndex(cnfItems, newTpSlice, true)
 			if err != nil {
-				return nil, nil, false, nil
+				return nil, nil, nil, false, nil
 			}
 			ranges := res.Ranges
 			accesses = res.AccessConds
 			filters = res.RemainedConds
 			if len(accesses) == 0 {
-				return FullRange(), nil, true, nil
+				return FullRange(), nil, nil, true, nil
 			}
 			if len(filters) > 0 {
 				hasResidual = true
 			}
 			totalRanges = append(totalRanges, ranges...)
 			newAccessItems = append(newAccessItems, expression.ComposeCNFCondition(d.sctx, accesses...))
+			if res.ColumnValues != nil {
+				if i == 0 {
+					columnValues = res.ColumnValues
+				} else {
+					// take the intersection of the two columnValues
+					for j, valInfo := range columnValues {
+						if valInfo == nil {
+							continue
+						}
+						sameVale, err := isSameValue(d.sctx.GetSessionVars().StmtCtx, valInfo, res.ColumnValues[j])
+						if err != nil {
+							return nil, nil, nil, false, errors.Trace(err)
+						}
+						if !sameVale {
+							columnValues[j] = nil
+						}
+					}
+				}
+			}
```

Comment:
Also, seems `sameVale` is a typo.

Reply:
Yes, it is simpler but it may miss some constant columns. Here is a case:
```
create table t(a int, b int, c int, d int, index idx_a_b_c_d(a, b, c, d));
select * from t where (a = 3 and b < 1 and c = 4) or (a = 3 and b > 5 and c = 4) order by b, d;
```
In the case, `totalRanges` is `[3 -inf,3 1), (3 5,3 +inf]` while both `a` and `c` are constant.